### ISSUE TITLE
BUG: array interface c-struct memory management, the wrong reference is held

### ISSUE THREAD

**Issue by burlen** [numpy/numpy#20673]
### Describe the issue:

This issue pertains to the c-struct portion of the numpy array interface[[1](https://numpy.org/doc/stable/reference/arrays.interface.html#c-struct-access)]. When numpy makes use of an object's data via the `object.__array_struct__` member, numpy should take a reference to the PyCapsule stored in `object.__array_struct__` [[2](https://numpy.org/doc/stable/reference/arrays.interface.html#object.__array_struct__)]. When numpy releases that reference, the PyCapsule's destructor [[3](https://docs.python.org/dev/c-api/capsule.html#c.PyCapsule_Destructor)] can free the data shared with numpy.

However, numpy takes a reference to the `object` itself[[4](https://github.com/numpy/numpy/blob/48deea3f616c263e262d9147d62db5ae6c284683/numpy/core/src/multiarray/ctors.c#L2108)] rather than the PyCapsule in `object.__array_struct__`, and immediately DECREF's the PyCapsule[[5](https://github.com/numpy/numpy/blob/48deea3f616c263e262d9147d62db5ae6c284683/numpy/core/src/multiarray/ctors.c#L2112)].   This approach is  fine when the data shared with numpy is a pointer to memory directly managed by the object, but is incorrect when a temporary is needed to share the data with numpy because when numpy DECREF's the PyCapsule, the PyCapsule destructor will free the temporary leading to invalid memory accesses when accessed from Python/numpy code.

Numpy should take a reference to the PyCapsule instance in `object.__array_struct__`  rather than to the `object` itself. The PyCapsule's destructor is the only means we have of numpy signaling that numpy is finished with the _specific data_ that was shared with it. When numpy DECREF's the `object` itself, on the C side we can't tell when Numpy is finished, and hence we can't know when to free the _specific data_ that was shared with numpy. This means that any temporary allocated for the purposes of sharing data with numpy needs to persist for the lifetime of the object itself, rather than only the time it is being used by numpy. That's not to hard to orchestrate but it is inefficient and problematic for large arrays.

For instance in my case I want to pass data to numpy that may have been allocated either on a GPU or CPU. In the case that it is allocated on a CPU, the pointer is indeed managed by the object, and numpy's approach of holding a reference to the object works. However, in the case that the data was allocated on a GPU, when sharing with numpy I first need to move the data from the GPU to a temporary allocation on the CPU for numpy. that temporary needs to persist while the corresponding numpy array is alive. this copy exists _only for numpy_, and should not outlive numpy's use of it. if I associate the temporary with the PyCapsule in `object.__array_struct__` there are invalid memory accesses, because numpy decref's the PyCapsule before the memory should be released. If I associate the temporary with the object itself then the temporary ends up staying alive long past numpy's use of it, until the object itself is destroyed. The latter could lead to out of memory conditions and is inefficient.

Holding a reference to both `object` and the PyCapsule in `object.__array_struct__` the would solve the problem and ensure that codes relying on the current implementation continue to function.

### Reproduce the code example:

```python
a small self contained example can be found here:
https://github.com/burlen/test_array_interface
```

### Error message:

```shell
see example above
```

### NumPy/Python version information:

1.19.5 3.9.0 (default, Oct  6 2020, 00:00:00) 
[GCC 10.2.1 20200826 (Red Hat 10.2.1-3)]

**Comment by seberg** [comment]
NumPy currently assumes that the original object owns the data.  I am wondeirng if we should actually change this.  Since you should be able to use `__array__` if you cannot provide a no-copy interface?

**Comment by burlen** [comment]
> you should be able to use `__array__` if you cannot provide a no-copy interface?

But isn't Numpy's array interface supposed to hold a reference to PyCapsule it requires I provide?

For instance the documentation [here](https://numpy.org/doc/stable/reference/arrays.interface.html#c-struct-access) says:

> `object.__array_struct__`
> 
> > A PyCapsule whose pointer member contains a pointer to a filled PyArrayInterface structure. Memory for the structure is dynamically created and the PyCapsule is also created with an appropriate destructor so the retriever of this attribute simply has to apply Py_DECREF to the object returned by this attribute when it is finished. Also, either the data needs to be copied out, or a reference to the object exposing this attribute must be held to ensure the data is not freed. Objects exposing the `__array_struct__` interface must also not reallocate their memory if other objects are referencing them.

> 

My reading of this is that Numpy will hold a reference to the PyCapsule until it is finished. To me this approach makes perfect sense since I provide the capsule's destructor I have full control over what references are held while Numpy is using the shared data. The approach is simple and could handle different scenarios ranging form the case where only a reference to the object providing the data need be held, to the case where references to other resources need to be held.

**Comment by seberg** [comment]
My reasoning is that I think the protocol should probably guarantee a no-copy operation (i.e. a view).  If we tie the data management to the capsule rather than the original object, this is not enforced anymore.  On the other hand `__array__` is a way to create a new object whatever way you like (with a copy).

**Comment by burlen** [comment]
I'm not sure why you'd need to enforce constraints, other than that the shared data stays alive, on the code sharing the data. Holding a reference to the capsule doesn't preclude no-copy data sharing, but it enables uses cases that can't be realized by no-copy. Holding an additional reference to the capsule has a very low overhead. It seems to me that the array interface protocol is documented as working that way.

On the other hand, I have not come across `__array__` before.  I'd like to understand the approach you are suggesting with `__array__` and see how I could apply it to my use case. In my use case I would like to use the array interface protocol to share data with numpy from a C++ code. I've wrapped the C++ data structure into Python using SWIG and exposed the data to numpy via the `__array_struct__` attribute as documented in the array interface protocol. Here is a small self contained example that I've used for testing: https://github.com/burlen/test_array_interface

Is [this](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.__array__.html) the `__array__` that  you are referring to? Given that it's a member of the ndarray class I'm not sure how it's applicable to this use case, and so I wonder if I'm looking at the right thing.

Perhaps you could point me to an example of sharing data from a C++ code via `__array__`, or show how my example would need to be modified?

**Comment by seberg** [comment]
>> Also, either the data needs to be copied out, or a reference to the object exposing this attribute must be held to ensure the data is not freed.

This documents that the object holds the data, not the capsule (the current docs are not contrary to the implementation).   The docs for `__array__` are unfortunate (it was bad, and at some point got even worse).  I do not have time to search for better examples right now, there should be many in other projects though, e.g. pandas, pytorch, etc.

I thought there was a new doc page on some of these, but I think that is probably still an open PR.  Additionally to array-interface, there is also the Python buffer interface you could use.  But in general I would prefer you first consider `__array__` if you are not doing a no-copy exposure.  Or is there another use-case?

**Comment by burlen** [comment]
> > > Also, either the data needs to be copied out, or a reference to the object exposing this attribute must be held to ensure the data is not freed.
> 
> This documents that the object holds the data, not the capsule

There I disagree with you, this only says that a reference should be held. This does not say that the capsule should not be the holder of the reference.

I think that one of the advantages of holding a reference to the capsule is flexibility in what resources can be managed. It would easy to hold a reference to the object itself, as well as any other resources that need to be managed, in the capsule. I don't see a down side to it because it has a low computational and memory overheads, and it would make memory management easier in a wider variety of situations.

It seems we may have to agree to disagree. I will look for an alternate solution. I won't use `__array__` without more information since, it's not immediately clear to me how it could be used here, and it seems to be documented as for another purpose. A quick scan of the numpy sources indicate that the same situation will likely arise with the buffer protocol since a reference is not held by numpy, so I think that's not an option here. With some contortions, I can think of at least one way to make it work though.

Thanks for the feedback!

**Comment by seberg** [comment]
This is the PR adding some documentation for it: https://github.com/numpy/numpy/pull/20185

It does make a difference, because we may want to add an option to ensure that no copy is made in `np.array()`, and if we do that we have to know whether the `__array_interface__` created a copy or not:  This is much easier if we know it did not, and currently, that is de-facto guaranteed.

I don't care about holding one more reference, but if the whole point is to allow exposure after copy, `__array__` seems like the more obvious way to do achieve that:  It is just a method that returns a new NumPy array!

**Comment by burlen** [comment]
>  I think the confusion/disagreement is because the docs are written from the perspective of the consumer who calls `__array_struct__` rather than from the perspective of the exporter.

Yes, and in our conversation here, we also might be talking about things each with the opposite perspective in mind.

>  we may want to add an option to ensure that no copy is made in np.array(), and if we do that we have to know whether the `__array_interface__` created a copy or not: This is much easier if we know it did not, and currently, that is de-facto guaranteed.

I believe that here you're speaking of guarantees about what happens inside Numpy. To clarify the temporary buffers I'm speaking of are all occurring in my code not in Numpy itself. But I would need a way to release the temporary when numpy is done so as not to create a leak or have it disappear too soon.

>  if the whole point is to allow exposure after copy

It's not quite that simple. I need to share data with numpy by zero-copy transfer when possible (that is preferred), but also to have the option to make a copy (external to numpy) when zero-copy sharing is not possible. The example I shared above focused on the latter. I need both path ways, with the decisions about how the data is shared being made at runtime.

> ` __array__` seems like the more obvious way to do achieve that

The PR you shared helped clarify the approach, and got me to the right places in the existing documentation, (thanks). I think I can make this work.

**Comment by seberg** [comment]
To be clear, I am not completely against allowing what you want.  But I am not clear that it is actually the best thing for you to have or for interaction between NumPy and your project in the longer term.

One thing that should be valid, is to raise an `AttributeError` when getting `__array_struct__` (or `__array_interface__`) to defer to `__array__` only when a copy is necessary (since a copy is necessary, I assume that overhead is negligible anyway).
That way you can use `__array_struct__` for the CPU backed data.

> I believe that here you're speaking of guarantees about what happens inside Numpy. 

The idea is that a user could say `np.array(your_obj, copy=NEVER)` and it should (ideally) succeed if it is a CPU backed object and fail if it is GPU backed (if it is CPU backed, the result is a view into the original data, if it is GPU backed it is not).  Array-struct gives us no way to signal this, but right now we can assume that the export is no-copy.

(To be clear, to really support this well, `__array__` should get a new argument, and if the implementer does not accept it, we would assume that a copy is made.)

**Comment by mattip** [comment]
> I think the protocol should probably guarantee a no-copy operation (i.e. a view).

- Is the no-copy operation specified somewhere?
- How does it conflict with holding a reference to the PyCapsule instead of to the object inside the PyCapsule?
- I think we should deprecate `__array_struct__` in the long run. It offers no advantage over dlpack (or `memoryview` for more complicated dtypes), so in any case this is a short-term fix.

**Comment by seberg** [comment]
> Is the no-copy operation specified somewhere?

Probably not, and maybe should be clarified if we would like it to be true.  But since the original object must be the owner of the data I somewhat expect it is true in practice. (Since the owner could copy during export, but that copy has to be persistent to the user).

> How does it conflict with holding a reference to the PyCapsule instead of to the object inside the PyCapsule?

It doesn't conflict, but if we can assume the data is always owned by the original object, there is no reason for holding on to the capsule anymore?

> I think we should deprecate __array_struct__ in the long run. It offers no advantage over dlpack (or memoryview for more complicated dtypes), so in any case this is a short-term fix.

Right, and little advantage over `__array_interface__` except being a bit faster because it is not a dict. But at least `__array_interface__` could probably extended to export a `dtype` object rather than strings to support user DTypes (which would be hard for `__array_struct__`) – which is a possible advantage over those protocols.  OTOH, we do have the `__array__` protocol as well, which is maybe easier to extend.

**Comment by burlen** [comment]
Thanks for the hint about raising AttributeError. Also very good to know about the plans to deprecate `__array_struct__`. I'm still prototyping so this is timely information and I'll make use of `__array_interface__` instead. I'm also very happy to have learned about `__array__` method and get exposure to the more modern methods being implemented for sharing data between Python libraries.

I'm hung up on a couple things with respect to the line of reasoning around `copy=NEVER` as it's being applied to the data shared with numpy from an external source above.

Mentioned above in a few places is that the current implementation guarantees that the user isn't providing Numpy a copy of data being shared. I don't think that this is guaranteed. Numpy is internally holding a reference to the object that provides the capsule via the `__array_struct__` attribute. The object sharing the data could in theory manage copies internally, which are then shared with Numpy. I didn't want to do it that way because my object can long outlive Python scripts that access it, and any temporary buffers allocated for the purposes of sharing with Numpy would then persist longer than necessary and I worried that it could potentially cause OOM conditions.

How do the `copy=NEVER`  arguments made above against Numpy also (or instead of) holding a reference to the PyCapsule instance, apply to Numpy's current implementation holding a reference to the object that provided it?

Note that the situation is similar for the `__array_interface__` in that Numpy holds a reference to the object that provides the `__array_interface__` dictionary.

**Comment by seberg** [comment]
You are right, the `copy=NEVER` is an argument only, and not a waterproof reason.  The functionality exists (but not publicly!), and right now it assumes that a copy is not made (which is indeed *not* quite guaranteed although but probably normally true in practice!).

I should have been more gentle probably, my point is mainly that I see this as an enhancement (not a bugfix) and that it seems like an unnecessary enhancement (due to `__array__` and other protocols existing).  And my gut reaction is that including it may make the story around `copy=NEVER` even trickier than it already is.

**Comment by burlen** [comment]
> I see this as an enhancement

That would be reasonable. I was on the fence. The reason I thought it might be a bug was that in a couple of spots the documentation seemed to indicate, at least to the inexperienced eye, that a reference to the capsule will be held. In addition to the spot([here](https://numpy.org/doc/stable/reference/arrays.interface.html#c-struct-access)) I pointed to above, there's the following  paragraph is describing the difference between the current version of the array interface protocol and the previous ([here](https://numpy.org/doc/stable/reference/arrays.interface.html#differences-with-array-interface-version-2))

> The context member of the PyCapsule (formally the desc member of the PyCObject) returned from `__array_struct__` was not specified. Usually, it was the object exposing the array (so that a reference to it could be kept and destroyed when the C-object was destroyed). It is now an explicit requirement that this field be used in some way to hold a reference to the owning object.

It's that last sentence that made me think that a reference to the capsule would be held while Numpy needed access to the shared data, and that it was up to me hold a reference to the object sharing the data in the capsule. It's the wording that made me think it worked that way once and was a regression.

I also found [this](http://blog.audio-tk.com/2008/11/04/exposing-an-array-interface-with-swig-for-a-cc-structure/) example from 2008 illustrating use of version 2 of the array interface. The author had used the `desc` slot in the Py CObject to hold a reference to his data structure.

Even if it's not a bug, it's not a crazy idea. For instance the Numba CUDA array interface made provisions for holding either a reference to the object providing the data, or some other object ([here](https://numba.pydata.org/numba-doc/dev/cuda/cuda_array_interface.html#lifetime-management-in-numba)).

> my gut reaction is that including it may make the story around copy=NEVER even trickier than it already is.

Isn't `copy=NEVER` is a policy about what Numpy does with the data once it has it? Is it not the case that, regardless of the source of the data being shared with Numpy, Numpy could error out if it needed to copy the data?

Who tells Numpy `copy=NEVER`? Is it the sharer of the data? If so a flag in the PyArrayInterface [flags](https://numpy.org/doc/stable/reference/c-api/types-and-structures.html#c.PyArrayInterface.flags) field might be used?

**Comment by seberg** [comment]
TL;DR: You convinced me that it would make a lot of sense to modify the code to use `context != NULL ? context : exporting_obj` as the base for the new array.  (I.e. close to what you got, but slightly different.)

Thanks for digging in so deeply, clearly there is some catching up I need to do with respect to what `__array_struct__` specifies (or means to specify) – it is (to pressimistic me) surprisingly nicely specified :)!
And I actually never used the `context` pointer, which is cool to know, since it could be very useful for other (similar) things as well!

> It is now an explicit requirement that this field be used in some way to hold a reference to the owning object.

I think this mainly ensures that the capsule remains valid even when it is not consumed directly (i.e. I can use `myobj.__array_struct__ = other_array.__array_struct__` and `other_array` will remain alive.

But you are right, this means we do have two pieces of information:
* We already require the "owner" object to be explicitly specified in some form
* We could argue that we usually know that no-copy was made (not 100% of course!) if the context is `NULL` (it should not be) or the object itself, and we would have the choice to reject that.

So I think my objections are not serious, because we should be future proof in the vast majority of cases, and all others have a path of being fixed up.

I also had not realized that we actually do have a form of versioning going on here (not ideal, but not bad).  I would agree that a flag could make sense, but I think a logic of "if context and exporter matches" would probably just good enough as an amendment to the protocol

**Comment by charris** [comment]
Was this fixed with the merge of the related PR?

**Comment by burlen** [comment]
yes
